ethcore/src/verification/queue/mod.rs

```diff
@@ -505,6 +506,16 @@ impl<K: Kind> VerificationQueue<K> {
 				match err {
 					// Don't mark future blocks as bad.
 					Error(ErrorKind::Block(BlockError::TemporarilyInvalid(_)), _) => {},
+					// If the transaction root or uncles hash is invalid, it doesn't necessarily mean
+					// that the header is invalid. We might have just received a malformed block body,
+					// so we shouldn't put the header hash to `bad`.
+					//
+					// We still put the entire `Item` hash to bad, so that we can early reject
+					// the items that are malformed.
+					Error(ErrorKind::Block(BlockError::InvalidTransactionsRoot(_)), _) |
+					Error(ErrorKind::Block(BlockError::InvalidUnclesHash(_)), _) => {
```

Comment:
```suggestion
					Error(ErrorKind::Block(BlockError::InvalidUnclesHash(_)), _) |
					Error(ErrorKind::Block(InvalidReceiptsRoot(_)), _) => {
```

Invalid receipt root probably also belong here.

Reply:
I don't think so. This happens after we execute transactions and that just means that given the `Header` with correct `transactions_root`, we got a different `receipts_root` (i.e. execution was different), which should definitely mark the entire header as invalid.